// @hedgedoc/markdown-it-plugins/dist/cjs/image-size/parse-image-size.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseImageSize = void 0;
const specialCharacters_js_1 = require("./specialCharacters.js");
function isCharacterADigit(code) {
    return code >= specialCharacters_js_1.SpecialCharacters.NUMBER_ZERO && code <= specialCharacters_js_1.SpecialCharacters.NUMBER_NINE;
}
function findNextNotNumberCharacter(startPosition, maximalPosition, content) {
    for (let position = startPosition; position < maximalPosition; position += 1) {
        const code = content.charCodeAt(position);
        if (!isCharacterADigit(code) && code !== specialCharacters_js_1.SpecialCharacters.PERCENTAGE) {
            return position;
        }
    }
    return maximalPosition;
}
function parseNextNumber(content, startPosition, maximalPosition) {
    const endCharacterIndex = findNextNotNumberCharacter(startPosition, maximalPosition, content);
    return {
        position: endCharacterIndex,
        value: content.slice(startPosition, endCharacterIndex)
    };
}
const checkImageSizeStart = (code) => {
    return (code === specialCharacters_js_1.SpecialCharacters.LOWER_CASE_X ||
        (code >= specialCharacters_js_1.SpecialCharacters.NUMBER_ZERO && code <= specialCharacters_js_1.SpecialCharacters.NUMBER_NINE));
};
function parseImageSize(imageSize, startCharacterPosition, maximalCharacterPosition) {
    if (startCharacterPosition >= maximalCharacterPosition) {
        return;
    }
    let currentCharacterPosition = startCharacterPosition;
    if (imageSize.charCodeAt(currentCharacterPosition) !== specialCharacters_js_1.SpecialCharacters.EQUALS) {
        return;
    }
    currentCharacterPosition += 1;
    if (!checkImageSizeStart(imageSize.charCodeAt(currentCharacterPosition))) {
        return;
    }
    const width = parseNextNumber(imageSize, currentCharacterPosition, maximalCharacterPosition);
    currentCharacterPosition = width.position;
    const code = imageSize.charCodeAt(currentCharacterPosition);
    if (code !== specialCharacters_js_1.SpecialCharacters.LOWER_CASE_X) {
        return;
    }
    currentCharacterPosition += 1;
    const height = parseNextNumber(imageSize, currentCharacterPosition, maximalCharacterPosition);
    currentCharacterPosition = height.position;
    return {
        width: width.value,
        height: height.value,
        position: currentCharacterPosition
    };
}
exports.parseImageSize = parseImageSize;
